### PR TITLE
Add support for struct field based filtering

### PR DESCRIPTION
**Pull request by prodeezy**
Addresses [Issue#122](https://github.com/apache/incubator-iceberg/issues/122) 

Update: 
This PR should be ready for full review. I'v added tests and comments where possible. Addressed most PR comments. 

A basic test run with this change run on pre-generated Iceberg formatted data containing struct level metrics (Run with Spark patched to push struct filters down to Iceberg)  : 

```
val schema = new StructType().add("age", IntegerType).add("name", StringType).add("friends", MapType(StringType, IntegerType)).add("location", new StructType().add("lat", DoubleType).add("lon", DoubleType))

val iceDf = spark.read.format("iceberg").load("iceberg-people-nestedfield-metrics")
iceDf.createOrReplaceTempView("iceberg_people_nestedfield_metrics")


// Struct filter pushed down by Spark to Iceberg Scan
scala> spark.sql("select * from iceberg_people_nestedfield_metrics where location.lat = 101.123").explain()
== Physical Plan ==
*(1) Project [age#0, name#1, friends#2, location#3]
+- *(1) Filter (isnotnull(location#3) && (location#3.lat = 101.123))
   +- *(1) ScanV2 iceberg[age#0, name#1, friends#2, location#3] (Filters: [isnotnull(location#3), (location#3.lat = 101.123)], Options: [path=iceberg-people-nestedfield-metrics,paths=[]])

// Without this PR following query would fail with the exception listed in Issue#22
scala> spark.sql("select * from iceberg_people_nestedfield_metrics where location.lat = 101.123").show()
+---+----+--------------------+-----------------+
|age|name|             friends|         location|
+---+----+--------------------+-----------------+
| 30|Andy|[Josh -> 10, Bisw...|[101.123, 50.324]|
+---+----+--------------------+-----------------+


scala> spark.sql("select * from iceberg_people_nestedfield_metrics where location.lat <= 101.123").show()
+---+----+--------------------+-----------------+
|age|name|             friends|         location|
+---+----+--------------------+-----------------+
| 30|Andy|[Josh -> 10, Bisw...|[101.123, 50.324]|
+---+----+--------------------+-----------------+


scala> spark.sql("select * from iceberg_people_nestedfield_metrics where location.lat >= 101.123").show()
+---+------+--------------------+-----------------+
|age|  name|             friends|         location|
+---+------+--------------------+-----------------+
| 30|  Andy|[Josh -> 10, Bisw...|[101.123, 50.324]|
| 19|Justin|[Kannan -> 75, Sa...|[175.926, 20.524]|
+---+------+--------------------+-----------------+


scala> spark.sql("select * from iceberg_people_nestedfield_metrics where location.lat > 101.123").show()

+---+------+--------------------+-----------------+
|age|  name|             friends|         location|
+---+------+--------------------+-----------------+
| 19|Justin|[Kannan -> 75, Sa...|[175.926, 20.524]|
+---+------+--------------------+-----------------+
```

Gist to create  data : https://gist.github.com/prodeezy/001cf155ff0675be7d307e9f842e1dac

/cc @rdblue @aokolnychyi  @xabriel  @fbocse

### CODE REVIEW

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java [hunk 1]
```diff
@@ -67,11 +68,13 @@ Expression bind(Types.StructType struct) {
...
     Types.NestedField field;
-    if (caseSensitive) {
-      field = struct.field(ref().name());
-    } else {
-      field = struct.caseInsensitiveField(ref().name());
-    }
+    String expressionFieldPath = ref().name();
+
+    boolean isStructFieldExp = expressionFieldPath.indexOf('.') > -1 &&
+      struct.field(expressionFieldPath.split("\\.")[0]).type() instanceof Types.StructType;
```
Comment (fbocse) [inline api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: Nit: in terms of where we want to put in the constraint of filtering only against struct types, I'm wondering if it'd make sense to only code this aspect in the private `findStructField` method.  

We'd only have `bind` delegate to `findStructField` (findNestedField ?)  in the case that the `expressionFieldPath` contains nested fields only so we'd simplify the condition as `boolean isFieldExp = expressionFieldPath.indexOf('.') > -1`.

I'm thinking that if we may end up adding support for other nested types (i.e. map) we'd make changes only to the `findStructField` implementation without having to change the internals of `bind` to delegate to a potential private method `findMapField` and adding cyclomatic complexity to evaluating the `field` attribute.

Wdyt?

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java [hunk 2]
```diff
@@ -32,19 +31,41 @@
...
+          Types.StructType subStruct = fields.get(i).type().asStructType();
+
+          Type ret = findStructFieldType(fieldId, subStruct);
+          if (ret != null) {
+            return ret;
+          }
+        }
+      }
+    }
+    return null;
```
Comment (fbocse) [inline api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: Should this method be returning a null value in case none of the nested fields of this struct match by field id?
Instead of an instance of `BoundReference` with a null private attribute `type` would it be acceptable that we instead `throw new ValidationException("Cannot find nested field id %d in struct: %s", fieldId, struct);` ?

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java [hunk 3]
```diff
@@ -32,19 +31,41 @@
 
   BoundReference(Types.StructType struct, int fieldId) {
     this.fieldId = fieldId;
-    this.pos = find(fieldId, struct);
-    this.type = struct.fields().get(pos).type();
+    this.pos = findTopFieldPos(fieldId, struct);
+    this.type = (pos > -1) ? struct.fields().get(pos).type() : findStructFieldType(fieldId, struct);
```
Comment (fbocse) [inline api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: 👍

**Comment by fbocse** [comment]
Thank you @prodeezy for contributing this, it looks really helpful!

**Comment by fbocse** [comment]
General API design related question - how do we bubble up in the API layer the fact that we are supporting the struct field based filtering but not other nested field types filtering? I am not referring  to the implementation details of exposing this in the API, I am just curious what a client's expectations should be that the API exposes this new capacity in a coherent manner.

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java (hunk 1 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: I like the idea. adding it in.

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java (hunk 2 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: done.

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java (hunk 1 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: done.

**Comment by prodeezy** [comment]
> General API design related question - how do we bubble up in the API layer the fact that we are supporting the struct field based filtering but not other nested field types filtering? ..  I am just curious what a client's expectations should be that the API exposes this new capacity in a coherent manner.

@fbocse  Currently it would be implied. The only feedback the client gets right now is in the Physical plan, that this filter is pushed down to Iceberg Scan level and if the client inspects the Scan (using  iceTable.newScan().filter(structFilterExp).planFiles ) , should be able to notice the appropriate files being skipped. AFAIK, this is consistent with how top-level field based filter expressions currently communicate to the client.

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java [hunk 4]
```diff
@@ -32,19 +32,41 @@
...
+          Types.StructType subStruct = fields.get(i).type().asStructType();
+
+          Type ret = findStructFieldType(fieldId, subStruct);
+          if (ret != null) {
+            return ret;
+          }
+        }
+      }
+    }
+    throw new ValidationException("Cannot find nested field id %d in struct: %s", fieldId, struct);
```
Comment (rominparekh) [inline api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: minor: do we need to populate `nested` in the error message? My concern is that since we have now changed the functionality of `find` function (now returning `-1` instead of throwing an exception) it might make sense to drop the word `nested` altogether?

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java [hunk 5]
```diff
@@ -67,11 +68,12 @@ Expression bind(Types.StructType struct) {
...
-      field = struct.field(ref().name());
-    } else {
-      field = struct.caseInsensitiveField(ref().name());
-    }
+    String expressionFieldPath = ref().name();
+
+    boolean isNestedFieldExp = expressionFieldPath.indexOf('.') > -1;
+
+    field = isNestedFieldExp ? findNestedField(struct, expressionFieldPath, caseSensitive) :
+      caseSensitive ? struct.field(ref().name()) : struct.caseInsensitiveField(ref().name());
```
Comment (rominparekh) [inline api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: minor: can we reuse `expressionFieldPath` instead of obtaining `ref().name()` again?

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java [hunk 6]
```diff
@@ -32,19 +32,41 @@
...
+      if (fields.get(i).fieldId() == fieldId) {
+        return fields.get(i).type();
+      } else {
+
+        // look under struct by calling this method recursively
+        if (fields.get(i).type() instanceof Types.StructType) {
+          Types.StructType subStruct = fields.get(i).type().asStructType();
+
+          Type ret = findStructFieldType(fieldId, subStruct);
+          if (ret != null) {
```
Comment (xabriel) [inline api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: Minor: Looks like your recursive call would never return `null`.

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java [hunk 7]
```diff
@@ -132,4 +134,36 @@ public Expression bind(Types.StructType struct, boolean caseSensitive) {
...
+
+      i++;
+
+      nextFieldType = caseSensitive ? subField.field(nestedFields[i]).type() :
+        subField.caseInsensitiveField(nestedFields[i]).type();
+    }
+
+    return subField.field(lastFieldInPath);
+
+  }
```
Comment (xabriel) [inline api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: Now that we need to recurse, I wonder if it makes sense to reuse the indexes available in `TypeUtils`: https://github.com/apache/incubator-iceberg/blob/0c9c63140e838875dc8cc52a57be2c8f24ad9975/api/src/main/java/com/netflix/iceberg/types/TypeUtil.java#L78-L84 so that this code simplifies to:

```java
Integer idx = indexByName.get(expressionFieldPath);
Types.NestedField field = indexById.get(idx);
```
That code doesn't check for non-struct parents; we'd have to perhaps create a custom Indexer. Also we would need to calculate the index lazily and perhaps cache it ( As in the `Schema`, see https://github.com/apache/incubator-iceberg/blob/master/api/src/main/java/com/netflix/iceberg/Schema.java#L59-L71 )

Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: Doesn't look like TypeUtils.indexByName & TypeUtils.indexById are used when reading data. It's used by the Manifest reader to index on manifest file statistics data.

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java (hunk 4 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: done.

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java (hunk 5 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: done

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java (hunk 6 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: done

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java (hunk 7 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: I guess you are suggesting we now start using this. i dunno what impact that would have though. do you see a major benefit to it? I can look into it.

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java [hunk 8]
```diff
@@ -32,19 +32,41 @@
 
   BoundReference(Types.StructType struct, int fieldId) {
     this.fieldId = fieldId;
-    this.pos = find(fieldId, struct);
-    this.type = struct.fields().get(pos).type();
+    this.pos = findTopFieldPos(fieldId, struct);
```
[the remark this answers is not in the document]
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: Yeah, good point. going to write some unit tests to evaluate impact. will handle accordingly.

**Comment by prodeezy** [comment]
Added struct field based unit testing to TestDictionaryRowGroupFilter & TestMetricsRowGroupFilter.

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java (hunk 7 above)
Comment (xabriel) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: > I guess you are suggesting we now start using this.

Right.

> do you see a major benefit to it?

Code reuse. But it would only make sense if `findNestedField()` is called many times. The first call to `findNestedField()` would build the indexes and consult them. Then any subsequent call to `findNestedField()` would just consult the cached Maps.

Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: Right. seems like it's being called many times but by different expression evaluators. A cache to lookup field name to Types.NestedField would help. Good call.  will add one in.

**Comment by prodeezy** [comment]
@rdblue  thoughts?

**Comment by rdblue** [comment]
@prodeezy, thanks for working on this! I'll review it as soon as I get time this week.

**Comment by prodeezy** [comment]
@aokolnychyi   would be good to get your thoughts on this as well since you did all the upstream work to enable this :-)

**Comment by prodeezy** [comment]
Looks like `[pull#138](https://github.com/apache/incubator-iceberg/pull/138)` introduced conflicts. Will rebase  with latest on master and push.

**Comment by prodeezy** [comment]
> Looks like [pull#138](https://github.com/apache/incubator-iceberg/pull/138) introduced conflicts. Will rebase with latest on master and push.

done.

**Comment by rdblue** [comment]
@prodeezy, this is looking good. I think you have the right approach, but some things are missing.

With this patch, `BoundReference` has the correct type and field ID. That works for `InclusiveMetricsEvaluator` and the Parquet filters that look up data based on ID. But it isn't enough for other evaluators that require binding:

* `Evaluator` calls `BoundReference.get` to get a value from a `StructLike`. That method needs to be updated to handle struct nesting, preferably using an [`Accessor`](https://github.com/rdblue/iceberg/blob/support-predicates-on-nested-fields/api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java#L61-L217) (that's an unfinished branch of mine working on this problem). As it is now, evaluators would fail because `get` will try to access position -1 for any nested field.
* `InclusiveManifestEvaluator` uses `BoundReference.pos` to get the correct `PartitionFieldSummary` from a `ManifestFile` instance. The partition field summaries are stored in the order of partition fields, so binding to the partition struct gives the correct position in the field summaries list. I think this would actually work because all fields are top-level, but I think there should be a better signal than returning -1.

I think this should also bind differently. We never parse identifiers in Iceberg. Instead of parsing, we build indexes using the possible field names. That way, we never have to worry about quoting.

For example, if a user passes in an expression for "a.b.c", that could multiple paths: "a" / "b" / "c" or "a.b" / "c" or "a" / "b.c" or "a.b.c". But the important thing is that there can be only once column that flattens to "a.b.c" because the columns are ambiguous otherwise. Rather than searching, we keep a map in the schema from "a.b.c" to the right field ID.

Expression binding should use an index like this and then build accessors using the actual path of field names.

File: api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java [hunk 9]
```diff
@@ -67,29 +68,25 @@ Expression bind(Types.StructType struct) {
...
   public Expression bind(Types.StructType struct, boolean caseSensitive) {
-    Types.NestedField field;
-    String expressionFieldPath = ref().name();
 
-    boolean isNestedFieldExp = expressionFieldPath.indexOf('.') > -1;
-
-    field = isNestedFieldExp ? findNestedField(struct, expressionFieldPath, caseSensitive) :
-      caseSensitive ? struct.field(expressionFieldPath) : struct.caseInsensitiveField(ref().name());
+    Schema schema = new Schema(struct.fields());
+    Types.NestedField field = schema.findField(caseSensitive? ref().name(): ref().name().toLowerCase());
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java]: This should probably use a case insensitive version of `findField` instead of assuming that passing lower-case in will work.

Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java]: will do.

**Comment by prodeezy** [comment]
@rdblue  Is there a way to test the Evaluator as an end to end test? I haven't been able to test that part out properly with my conventional gist examples.

**Comment by rdblue** [comment]
@prodeezy, `IcebergGenerics` is a way to read a table directly and will use `Evaluator` to filter records. Just make sure that should updates unit tests for the evaluators as well as adding end-to-end tests.

**Comment by prodeezy** [comment]
Added tests for Evaluator and addressed other review comments.

File: api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/BoundReference.java]: handled this using accessors

File: api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java (hunk 7 above)
Comment (prodeezy) [reply api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java]: @xabriel I'm now creating an index that keeps ids to accessors and then use accessors to reach type and fields. access to this index is lazy as well.

**Comment by prodeezy** [comment]
@rdblue Thanks for taking a detailed look! I think i'v addressed the review comments and added pending tests.

**Comment by prodeezy** [comment]
@rdblue please take a look when you get a chance. thanks!

**Comment by rdblue** [comment]
@prodeezy, sorry for the delay and thanks for reminding me.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java [hunk 10]
```diff
@@ -19,57 +19,238 @@
...
   public int pos() {
+    if (pos == -1) {
+      throw new ValidationException("Cannot find position for non-primitive field id %d", fieldId);
+    }
     return pos;
   }
 
   public T get(StructLike struct) {
-    return struct.get(pos, javaType());
+    return (T) accessor.get(struct);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: I think this cast requires `@SuppressWarnings("unchecked")` right?

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java [hunk 11]
```diff
@@ -19,57 +19,238 @@
...
   }
 
   @Override
   public String toString() {
-    return String.format("ref(id=%d, pos=%d, type=%s)", fieldId, pos, type);
+    return String.format("ref(id=%d, accessor=%s)", fieldId, accessor);
+  }
+
+  private interface Accessor<T> extends Serializable {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: I think it would be better to add the accessor code in a package-private class. No need to embed all of it here.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java [hunk 12]
```diff
@@ -19,57 +19,238 @@
...
+
+    @Override
+    public Type type() {
+      return accessor.type();
+    }
+  }
+
+  private transient Map<Integer, Accessor<StructLike>> idToAccessor = null;
+
+  private Map<Integer, Accessor<StructLike>> lazyIdToAccessor(Schema schema) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: The lazy pattern doesn't make sense here where a `BoundReference` is for just one predicate. There's no need to get accessors for the whole schema or keep the result around.

Maybe accessors should be provided by the `Schema`? Or maybe this should be in the `Binder`. That would keep `Schema` clean and reuse the same accessors for the schema.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java (hunk 10 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: Also, is it possible to get rid of this cast by adding a return type to the accessor?

File: api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java [hunk 13]
```diff
@@ -66,28 +68,27 @@ Expression bind(Types.StructType struct) {
    * @throws ValidationException if literals do not match bound references, or if comparison on expression is invalid
    */
   public Expression bind(Types.StructType struct, boolean caseSensitive) {
-    Types.NestedField field;
-    if (caseSensitive) {
-      field = struct.field(ref().name());
-    } else {
-      field = struct.caseInsensitiveField(ref().name());
-    }
+
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java]: Nit: please remove this unnecessary whitespace.

File: api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java [hunk 14]
```diff
@@ -130,6 +131,7 @@ public Expression bind(Types.StructType struct, boolean caseSensitive) {
 //          break;
       }
     }
-    return new BoundPredicate<>(op(), new BoundReference<>(struct, field.fieldId()), lit);
+    return new BoundPredicate<>(op(), new BoundReference<>(schema, field.fieldId()), lit);
   }
+
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java]: Nit: non-functional change.

File: api/src/test/java/org/apache/iceberg/expressions/TestEvaluatior.java [hunk 15]
```diff
@@ -47,14 +47,38 @@
   private static final StructType STRUCT = StructType.of(
       required(13, "x", Types.IntegerType.get()),
       required(14, "y", Types.IntegerType.get()),
-      optional(15, "z", Types.IntegerType.get())
+      optional(15, "z", Types.IntegerType.get()),
+      optional(16, "s1", Types.StructType.of(
+              Types.NestedField.required(17, "s2", Types.StructType.of(
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestEvaluatior.java]: Nit: indentation is incorrect. Continuing indents should be 4 spaces.

File: parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java [hunk 16]
```diff
@@ -22,16 +22,20 @@
 import java.io.File;
 import java.io.IOException;
 import java.util.UUID;
+
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java]: Nit: there should be no separations in import lists.

File: parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java [hunk 17]
```diff
@@ -68,17 +75,22 @@
       optional(5, "some_nulls", StringType.get()),
       optional(6, "no_nulls", StringType.get()),
       optional(7, "non_dict", StringType.get()),
-      optional(8, "not_in_file", FloatType.get())
+      optional(8, "struct_not_null", structFieldType),
+      optional(10, "not_in_file", FloatType.get())
+
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestDictionaryRowGroupFilter.java]: Nit: please remove this blank line in the middle of an expression.

File: api/src/main/java/org/apache/iceberg/Schema.java [hunk 18]
```diff
@@ -157,6 +157,16 @@ public Type findType(int id) {
     return null;
   }
 
+
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/Schema.java]: Nit: please don't add extra blank lines.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java [hunk 19]
```diff
@@ -19,57 +19,238 @@
 
 package org.apache.iceberg.expressions;
 
+import java.io.Serializable;
 import java.util.List;
+import java.util.Map;
+
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: Nit: there should be no separations in imports.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java [hunk 20]
```diff
@@ -19,57 +19,238 @@
...
-    return type;
+    return accessor.type();
   }
 
   public int fieldId() {
     return fieldId;
   }
 
   public int pos() {
+    if (pos == -1) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: It seems strange to me to throw a `ValidationException` in some cases. It is likely that code paths will not know whether the bound expression is for `a` or `a.b`, but the latter will fail here, during evaluation instead of during binding.

Could the evaluator that uses pos be updated so that this accessor could be used? Maybe PositionAccessor could support accessing summaries from List<PartitionFieldSummary>. That would be cleaner.

**Comment by rdblue** [comment]
@prodeezy, are you still working on this? I'd like to get it in before we release.

**Comment by prodeezy** [comment]
@rdblue sorry bout the delay. I'l have the pending comments addressed in a day or two.

**Comment by rdblue** [comment]
No problem! Just let me know if you'd like me to help out.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java (hunk 11 above)
Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: done.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java (hunk 10 above)
Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: added the annotation. adding a return type fails in Evaluator and ResidualEvaluator.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java (hunk 12 above)
Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: So i tried putting this into `Binder` but that is not the only place predicates are  bound seems like.  `UnboundPredicate.bind()` is called from `Projections`, `ResidualEvaluator`,  and `Binder` .  I also tried putting it in `UnboundPredicate` but that is not useful coz there's one created for every predicate. 

Seems like Schema  is a good place? So i'm putting it there. Lemme know if this looks any better to you.

Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: Side effect of this is `Accessor`  interface had to be made public

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java (hunk 20 above)
Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: Need some clarification here .. The evaluator in question is the `InclusiveManifestEvaluator` which evaluates partition fields for matching manifests. So the `ref.pos()` is going to be on partition fields. Afaik partition stats are kept separately in snapshot files. Is the partition source id used always to reference that field in schema? Can I assume this for logical partitions as well?  

e.g.  
```
"partitions": {
    "array": [
      {
        "contains_null": true,
        "lower_bound": {
          "bytes": "\u0013\u0000\u0000\u0000"
        },
        "upper_bound": {
          "bytes": "\u001e\u0000\u0000\u0000"
        }
      }
    ]
  }
```

**Comment by prodeezy** [comment]
@rdblue addressed all but one comment. Need some clarification/guidance on evaluating position for the bound exp reference.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java (hunk 20 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: I think you're saying that this is technically safe, and that's correct. The only code path that calls this that evaluator and it is always binding to a flat partition structure. That's why it can use the position: it knows that the array of partition summaries is in the same order as a tuple of partition values.

My point here is that it is brittle to bind to a struct type and use the position for something else, and also that it is a bad API to expose the position when no normal path uses the position directly. Instead, maybe that evaluator should get this position from the first accessor. That way, it validates that the partition field is not nested (should be a single position accessor). My original thought was to add a method to the accessor that can return one of the partition summaries from a list. That would work, too, but requires another accessor method so it isn't a great idea.

Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: Thanks for clarifying. So i'm now using `PositionAccessor` to fetch position during manifest evaluation. Throwing an error if it's a different kind of accessor.

Comment (prodeezy) [reply api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: done

**Comment by prodeezy** [comment]
I took another glance. Think i'v addressed all pending comments.  @rdblue

**Comment by prodeezy** [comment]
@rdblue gentle reminder. Lemme know if this looks ok to you.

**Comment by rdblue** [comment]
Thanks for the reminder. I should have time to review it this week. Sorry for the delay.

File: api/src/main/java/org/apache/iceberg/expressions/BoundReference.java [hunk 21]
```diff
@@ -19,57 +19,35 @@
...
+  @SuppressWarnings("unchecked")
   public T get(StructLike struct) {
-    return struct.get(pos, javaType());
+    return (T) accessor.get(struct);
   }
 
   @Override
   public String toString() {
-    return String.format("ref(id=%d, pos=%d, type=%s)", fieldId, pos, type);
+    return String.format("ref(id=%d, accessor=%s)", fieldId, accessor);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/expressions/BoundReference.java]: Accessors don't implement `toString`.

**Comment by rdblue** [comment]
@prodeezy, I thought that it would be easier if I made a few minor changes since it would take longer to ask you to do them than to just move a few things around. I opened a PR against your branch: https://github.com/prodeezy/incubator-iceberg/pull/1

If you agree with those changes, just merge the PR and push and I'll commit this. Thanks!

**Comment by rdblue** [comment]
@prodeezy, the test failures are because I removed the map column from `TestMetricsRowGroupFilter`. I thought that wasn't used but I guess it was. If you add that column back, the tests should pass.

**Comment by prodeezy** [comment]
Fixed test.

**Comment by rdblue** [comment]
Merged! Thanks for fixing this @prodeezy!
